pyk/src/pyk/proof/proof.py

```diff
@@ -40,7 +40,7 @@ class Proof(Generic[PS, SR]):
     :param SR: Step result: data produced by executing a PS with `Prover.step_proof` used to update the `Proof`
     """
 
-    _PROOF_TYPES: Final = {'APRProof', 'EqualityProof', 'RefutationProof'}
+    _PROOF_TYPES: Final = {'APRProof', 'EqualityProof', 'RefutationProof', 'MultiProof'}
```

Comment:
(forming a thread)

I would like to make the use case very precise. Imagine that we did have a branching `setUp`, with all of its final states, together with the proofs of the associated test for each of the branches. How are these connected in the `MultiProof`? How do I know which proof corresponds to which branch? Could you give an example - imagine `setUp` branches into 3 possibilities.

Also, from the test below, I see that `subproof_ids` are used, but I don't see a proof that's their superproof.

Reply:
I was thinking the constructor proof, setUp proof(s), and the main proof(s) could just go directly under one `MultiProof`.

To know which proof corresponds to each branch, we would probably set the proof ID of the test function subproofs to something which included the node ID of the setUp KCFG it copies its starting configuration from. So if the function is called `my_test`, and the setUp function branches into 3 and ends on nodes 10, 11, and 12, the subproof graph could look something like:
```
[MyContract.my_test (MultiProof)]-->(
    [MyContract.setUp (APRProof)], 
    [MyContract.my_test_setup_config_10 (APRProof)], 
    [MyContract.my_test_setup_config_11 (APRProof)], 
    [MyContract.my_test_setup_config_12 (APRProof)], 
)
```
And of course you can also inspect the initial configurations of the subproofs.

Reply:
Thoughts, @tothtamas28, @ehildenb?

Reply:
I guess the `subproof` abstraction cannot be used here because in `kontrol` these proof obligations are manually constructed from one another. i.e.

```
init -> constructor done
constructor done -> setup done
setup done -> target
------------------------------
init -> target
```

is done by hand (i.e. not by the prover backend).

I wonder, would it be possible to use subproofs if we changed the semantics? Because in a semantics where constructor and setup steps are handled internally, this sequential compositon can be handled by the backend.

Reply:
Example to better highlight the difference between subproofs and `MultiProof` as I understand it. Suppose we want to prove `a -> c`.

Subproof approach:

```
claim: a -> c
sub-claim: a -> b
```

If the proof passes, we know `a -> c`. The sub-claim might make proving more efficient, or might even be necessary for `a -> c` to pass.

`MultiProof` approach:

```
sub-claim: b -> c
sub-claim: a -> b
```

If the `MultiProof` passes, we know `a -> c`, but just because we constructed it so. The composition or the original calim `a -> c` is never machine-checked.